File: backend/src/main/java/com/example/demo/controller/OrderController.java
package com.example.demo.controller;

import com.example.demo.dto.request.CheckoutRequest;
import com.example.demo.dto.response.OrderResponse;
import com.example.demo.response.ApiResponse;
import com.example.demo.response.PageResponse;
import com.example.demo.service.interfaces.OrderService;
import jakarta.validation.Valid;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.web.PageableDefault;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

/** Module 8 — Checkout and orders. All endpoints require a logged-in user. */
@RestController
@RequestMapping("/api/orders")
public class OrderController {

    private final OrderService orderService;

    public OrderController(OrderService orderService) {
        this.orderService = orderService;
    }

    /**
     * POST /api/orders/checkout-preview
     * Calculates totals, savings and points WITHOUT saving anything.
     */
    @PostMapping("/checkout-preview")
    public ResponseEntity<ApiResponse<OrderResponse>> checkoutPreview(
            @Valid @RequestBody CheckoutRequest request) {
        return ResponseEntity.ok(ApiResponse.success(
                "Checkout preview calculated", orderService.checkoutPreview(request)));
    }

    /** POST /api/orders — converts the cart into a real order. */
    @PostMapping
    public ResponseEntity<ApiResponse<OrderResponse>> placeOrder(
            @Valid @RequestBody CheckoutRequest request) {
        return ResponseEntity.status(HttpStatus.CREATED).body(ApiResponse.success(
                "Order placed successfully. Proceed to payment.",
                orderService.placeOrder(request)));
    }

    /** GET /api/orders?page=0&size=10 */
    @GetMapping
    public ResponseEntity<ApiResponse<PageResponse<OrderResponse>>> getMyOrders(
            @PageableDefault(size = 10, sort = "orderDate", direction = Sort.Direction.DESC)
            Pageable pageable) {
        return ResponseEntity.ok(ApiResponse.success(
                "Orders retrieved successfully", orderService.getMyOrders(pageable)));
    }

    @GetMapping("/{orderId}")
    public ResponseEntity<ApiResponse<OrderResponse>> getOrder(@PathVariable Integer orderId) {
        return ResponseEntity.ok(ApiResponse.success(
                "Order retrieved successfully", orderService.getOrder(orderId)));
    }

    @PutMapping("/{orderId}/cancel")
    public ResponseEntity<ApiResponse<OrderResponse>> cancelOrder(@PathVariable Integer orderId) {
        return ResponseEntity.ok(ApiResponse.success(
                "Order cancelled", orderService.cancelOrder(orderId)));
    }

    /**
     * GET /api/orders/{orderId}/invoice-pdf
     *
     * Returns the PDF bytes directly rather than the ApiResponse envelope —
     * a browser or Postman needs the raw file to render or download it.
     */
    @GetMapping(value = "/{orderId}/invoice-pdf", produces = MediaType.APPLICATION_PDF_VALUE)
    public ResponseEntity<byte[]> downloadInvoice(@PathVariable Integer orderId) {

        byte[] pdf = orderService.generateInvoicePdf(orderId);

        return ResponseEntity.ok()
                .contentType(MediaType.APPLICATION_PDF)
                .header(HttpHeaders.CONTENT_DISPOSITION,
                        "attachment; filename=\"invoice-" + orderId + ".pdf\"")
                .body(pdf);
    }
}

File: dotnet-backend/Emart.Api/Controllers/OrderController.cs
using System.Threading.Tasks;
using Emart.Api.DTOs;
using Emart.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Emart.Api.Controllers
{
    // The route is spelled out rather than taken from [controller]: the class is
    // OrderController, which would have produced /api/order, while the client
    // and the API reference both use /api/orders.
    [ApiController]
    [Route("api/orders")]
    [Authorize]
    public class OrderController : ControllerBase
    {
        private readonly IOrderService _orderService;

        public OrderController(IOrderService orderService)
        {
            _orderService = orderService;
        }

        [HttpPost("checkout-preview")]
        public async Task<ActionResult<ApiResponse<OrderDto>>> CheckoutPreview([FromBody] CheckoutRequest request)
        {
            var preview = await _orderService.CheckoutPreviewAsync(request);
            return Ok(ApiResponse<OrderDto>.CreateSuccess("Checkout preview calculated", preview));
        }

        [HttpPost]
        public async Task<ActionResult<ApiResponse<OrderDto>>> PlaceOrder([FromBody] CheckoutRequest request)
        {
            var order = await _orderService.PlaceOrderAsync(request);
            return StatusCode(201, ApiResponse<OrderDto>.CreateSuccess(
                "Order placed successfully. Proceed to payment.", order));
        }

        [HttpGet]
        public async Task<ActionResult<ApiResponse<PageResponse<OrderDto>>>> GetMyOrders(
            [FromQuery] int page = 0, [FromQuery] int size = 10)
        {
            var orders = await _orderService.GetMyOrdersAsync(page, size);
            return Ok(ApiResponse<PageResponse<OrderDto>>.CreateSuccess("Orders retrieved successfully", orders));
        }

        [HttpGet("{orderId:int}")]
        public async Task<ActionResult<ApiResponse<OrderDto>>> GetOrder(int orderId)
        {
            var order = await _orderService.GetOrderAsync(orderId);
            return Ok(ApiResponse<OrderDto>.CreateSuccess("Order retrieved successfully", order));
        }

        [HttpPut("{orderId:int}/cancel")]
        public async Task<ActionResult<ApiResponse<OrderDto>>> CancelOrder(int orderId)
        {
            var order = await _orderService.CancelOrderAsync(orderId);
            return Ok(ApiResponse<OrderDto>.CreateSuccess("Order cancelled", order));
        }

        /// <summary>
        /// The ONE endpoint that does not return the JSON envelope: the browser
        /// needs raw PDF bytes to render or download the file.
        /// </summary>
        [HttpGet("{orderId:int}/invoice-pdf")]
        public async Task<IActionResult> DownloadInvoice(int orderId)
        {
            var pdf = await _orderService.GenerateInvoicePdfAsync(orderId);
            return File(pdf, "application/pdf", $"invoice-{orderId}.pdf");
        }
    }
}
